### DISCUSSION TITLE
How to change directory for x_cache/ and x_files/ in Quarto Website?

### DISCUSSION THREAD

**Discussion by jmgirard** [quarto-dev/quarto-cli#5184] [Q&A] [answered]
I have a quarto website where each main qmd file generates a cache and files directory. This ends up cluttering my main directory. Can I tell quarto (or knitr) to save generate these files in a subdirectory like _cache/?

**Comment by cscheid** [comment, marked as answer]
That's currently not supported by quarto, unfortunately. We're slowly fixing internal infrastructure issues that will make that easier in the future, but a proper fix for this is quite a ways out yet.

If you are using a quarto website, though, all of those directories should be going in the output of the project (rather than next to the .qmd files), so maybe I'm missing something.

**Reply by jmgirard** [reply]
Thanks for the speedy reply.

**Reply by jmgirard** [reply]
The `.html` files go to `/docs/` along with a copy of `_files/` but the `_cache/` and `_files/` directories remain next to the `.qmd`.

**Reply by mcanouil** [reply]
If you are really annoyed by this, you could create a post render script that cleanup your repository for whatever feel useless to you, see <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>.

**Reply by cscheid** [reply]
> The `.html` files go to `/docs/` along with a copy of `_files/` but the `_cache/` and `_files/` directories remain next to the `.qmd`.

I'm surprised by this, and can't reproduce it locally with a simple website project or a more complex one (namely, `quarto-web`). Can you share a repository that you experience this issue with?

**Reply by jmgirard** [reply]
I can try to make a MWE later but here is the real example:
https://github.com/jmgirard/film-dataset
_clip_info.qmd is a child document that gets called by many other qmds (e.g., AkeelahBee.qmd). One of its chunks has cache on.

**Reply by mcanouil** [reply]
I tried with a small setup, as expected `_files` gets moved to `output-dir` and as expected `_cache` remains in the working directory.
If your `_files` directory is not moved, it's likely because something did not go well or did not run completely (error, job aborted, etc.)
The `_cache` has no real reason to be in the output directory, it is only useful for computation.
This being said, `knitr` cache default directory can be changed at a chunk level or globally.

```html
knitr:
  opts_chunk:
    cache.path: _output/cache/
```
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/231890818-dc6e2f74-a28d-499e-a16f-0d7f6a7a5593.png">

Within a code cell, for example:

```{r}
#| cache: true
#| cache.path: !expr file.path("_output", sub("\\.rmarkdown", "_cache/", knitr::current_input()))
Sys.sleep(10)
```

Note that you can't evaluate code in global option as in code cell for some reason.
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/231892668-a5b32b1b-446f-4046-8e24-fa26fd874b7f.png">

**Reply by cscheid** [reply]
This is starting to feel like a bug in quarto for me, where I think you're expecting your `_*.qmd` files to be used only in the context of include targets, but they're both getting rendered and their outputs are not being cleaned up properly...

Unfortunately I can't get your website to compile locally because `renv::restore()` fails for me with

```
Warning: failed to find binary for 'StanHeaders 2.26.13' in package repositories
Warning: failed to find source for 'StanHeaders 2.26.13' in package repositories
```

**Reply by mcanouil** [reply]
Does the `_cache` directory (from `knitr`) is tied up to `_files` in Quarto? At least Quarto is changed the cache path to include file basename into it.

In my small and simple test, the cache is not cleaned up (as expected) but the `_files` is.

Also the repository is not clean, for example there is the `_site` directory while Quarto configuration file states the output is in `docs`. There are also some html files at the root of the repository.
Right now from the repository, it's hard to know where could be the issue.

**Reply by cderv** [reply]
Not going into the details around tweaking `cache.path` options but some context are documented in https://bookdown.org/yihui/rmarkdown-cookbook/cache-path.html#cache-path 

`rmarkdown::render()` should do the right thing regarding naming, no need to tweak it I believe, unless one want a global cache to speed up rendering. However, maybe Quarto is interfering with this as it handles cache too (https://quarto.org/docs/projects/code-execution.html#cache)  

> Does the _cache directory (from knitr) is tied up to _files in Quarto? 

In R Markdown and **knitr** they are both linked because when a cache chunk contains plots for example, the files folder is required. SOme context in https://github.com/rstudio/rmarkdown/issues/2408

I'll look into this closer to see how this all works within Quarto.

**Reply by cderv** [reply]
So I spent some time trying to understand how this all works between R Markdown, Knitr and Quarto

> The .html files go to /docs/ along with a copy of _files/ but the _cache/ and _files/ directories remain next to the .qmd.

About this, I think this is directly related to https://github.com/rstudio/rmarkdown/issues/2408 which we could improve. `*_files` is required for caching, so it is not part of the removed file. It is not returned by `rmarkdown::render` as part of `intermediates_dir`. This means Quarto will only copy it to output dir, but not _move_ it. If it were removed the caching would not work because the `_files` folder would not be recreated.  

So Quarto does copy the `*_files` dir as part of the `resourceFiles()` search when cache is activated, and will move the `*_files` dir when cache is not activated as part of `renderedFile.supporting` processing.  (which is a different one)

I believe this is the expected behavior currently for the cache to work. 

> I have a quarto website where each main qmd file generates a cache and files directory. This ends up cluttering my main directory.

This is indeed the behavior of **knitr**. We only have a `cache.path` options, and not a `cache.dir` options to change where the `cache.path` auto generated folder is saved. So by default it will be relative to the document.  
In **bookdown**, it does not clutter your root folder because we use a _trick_ to hide the `*_cache` and `*_files` folder in a global `_bookdown_files` folder , and then move them back to root during rendering. 
We don't have such behavior in Quarto. 

I don't think changing `cache.path` is really supported by **rmarkdown**. The `*_files ` folder will not be correctly handled. Maybe an issue in **rmarkdown** or a rather a feature request to handle cache configure better. 

So for now, there is not really a good workaround for the cluttering @jmgirard - except doing something with `pre` and `post` render script to do like bookdown maybe. But as @cscheid said we'll probably improve all this in the future. 

> This is starting to feel like a bug in quarto for me, where I think you're expecting your _*.qmd files to be used only in the context of include targets, but they're both getting rendered and their outputs are not being cleaned up properly...

@cscheid I don't think there is an issue, or I missed it. happy to chat live about this.

**Reply by cscheid** [reply]
> @cscheid I don't think there is an issue, or I missed it. happy to chat live about this.

You've looked at it deeper than I have, and I think your explanation makes sense; I was suspicious of us accidentally rendering the files that are used as include targets (because they have a `.qmd` extension), but I don't think that's the issue. I trust your expertise on this one.

**Reply by cderv** [reply]
> I was suspicious of us accidentally rendering the files that are used as include targets (because they have a .qmd extension)

Yes we're good on that part. The included file are rendered once included and caching + resources are correctly computed.
